connectors/src/connectors/notion/temporal/workflows/garbage_collection.ts

```diff
@@ -0,0 +1,297 @@
+        await logMaxSearchPageIndexReached({
+          connectorId,
+          searchPageIndex: pageIndex,
+          maxSearchPageIndex: MAX_SEARCH_PAGE_GARBAGE_COLLECTION_INDEX,
+        });
+        break;
+      }
+    } while (cursors[cursorType].last);
+
+    return { isComplete: true, pageIndex };
+  }
+
+  const [pagesResult, databasesResult] = await Promise.all([
+    runSearch("pages"),
+    runSearch("databases"),
+  ]);
+
+  const isComplete = pagesResult.isComplete && databasesResult.isComplete;
+
+  if (!isComplete) {
+    await Promise.all(promises);
+    await continueAsNew<typeof notionGarbageCollectionWorkflow>({
+      connectorId,
+      cursors,
+      pageIndexes: {
+        pages: pagesResult.pageIndex,
+        databases: databasesResult.pageIndex,
+      },
+    });
+    return;
+  }
+
+  // wait for all child workflows to finish
+  await Promise.all(promises);
+
+  // These are resources (pages/DBs) that we didn't get from the search API but that are
+  // child/parent pages/DBs of other pages that we did get from the search API. We upsert those as
+  // well.
+  let discoveredResources: {
+    pageIds: string[];
+    databaseIds: string[];
+  } | null;
+  do {
+    discoveredResources = await getDiscoveredResourcesFromCache({
+      connectorId,
+      topLevelWorkflowId,
+    });
+    if (discoveredResources) {
+      await markAsSeenAndUpsertNewResources({
+        connectorId,
+        pageIds: discoveredResources.pageIds,
+        databaseIds: discoveredResources.databaseIds,
+        runTimestamp,
+        childWorkflowQueue,
+        topLevelWorkflowId,
+      });
+    }
+  } while (discoveredResources && PROCESS_ALL_DISCOVERED_RESOURCES);
+
+  // Look at pages and databases that were not visited in this run, check with the notion API if
+  // they were really deleted and delete them from the database if they were.
+  // Find the resources not seen in the GC run
+
+  // Create batches of resources to check, by chunk of 100
+  const nbOfBatches = await createResourcesNotSeenInGarbageCollectionRunBatches(
+    {
+      connectorId,
+      batchSize: 100,
+    }
+  );
+
+  // For each chunk, run a garbage collection activity
+  const queue = new PQueue({
+    concurrency: MAX_PENDING_GARBAGE_COLLECTION_ACTIVITIES,
+  });
+  const gbPromises: Promise<void>[] = [];
+  for (let batchIndex = 0; batchIndex < nbOfBatches; batchIndex++) {
+    gbPromises.push(
+      queue.add(async () =>
+        garbageCollectBatch({
+          connectorId,
+          runTimestamp,
+          batchIndex,
+        })
+      )
+    );
+  }
+
+  await Promise.all(gbPromises);
+
+  // Once done, clear all the redis keys used for garbage collection
+  await completeGarbageCollectionRun(connectorId, nbOfBatches);
+
+  await sleep(INTERVAL_BETWEEN_SYNCS_MS);
+
+  await continueAsNew<typeof notionGarbageCollectionWorkflow>({
+    connectorId,
+  });
+}
+
+async function markAsSeenAndUpsertNewResources({
```

Comment:
This is the only new thing here. Because the "mark as seen and return new stuff" part is specific for GC, I moved it out of the performUpserts function.